golang/pkg/rnr/tasks.go

```diff
@@ -28,10 +33,84 @@ func taskSchedState(pbt *pb.Task) TaskState {
 	return UNKNOWN
 }
 
+// TaskCallback is installed by higher-level construct, such as callback task or nested task.
+type TaskCallback func(context.Context, *Task)
+type StateUpdateCallback func(*pb.Task) *pb.Task
+
 // Task is a generic interface for pollable tasks
-type Task interface {
-	Poll()
-	Proto(updater func(*pb.Task)) *pb.Task
-	SetState(pb.TaskState)
-	GetChild(name string) Task
+type Task struct {
+	cb           TaskCallback
+	pb           *pb.Task
+	children     []*Task
+	has_children bool
+}
+
+func NewTask(name string, children bool, cb TaskCallback) *Task {
+	return &Task{
+		cb: cb,
+		pb: &pb.Task{
+			Name:  name,
+			State: pb.TaskState_PENDING,
+		},
+		children:     []*Task{},
+		has_children: children,
+	}
+}
+
+func (task *Task) Poll() {
+	task.cb(context.TODO(), task)
+}
```

Comment:
Given that this PR already breaks lots of things, why not making this context aware:

```suggestion
func (task *Task) Poll(ctx context.Context) {
	task.cb(ctx, task)
}
```